endpoint rejestracji

wojbog opened the pull request:
dodanie ednpoint,walidacja danych,połączenie z bazą danych

wojbog committed:
endpoint rejestracji

dodanie ednpoint,walidacja danych,połączenie z bazą danych

shhhQuiettt commented on server.go:
```diff
@@ -11,5 +12,7 @@ func main() {
 
+	app.Post("/newUser",user.SetUser)//endpoint dodawania uzytkownika do bazy
+
 	app.Listen(":3000")
```
Zmień na 3001 w końcu

akruszewski replied:
Ta wartosc powinna isc z enva.

shhhQuiettt commented on serviceDB/serviceDB.go:
```diff
@@ -0,0 +1,30 @@
+
+
+func InsertUser(user *person.Person ) error {//dodawanie użytkownika
+	if errv:=baza.DB.Collection("zadania").FindOne(context.Background(), bson.M{"email":user.Email}); errv !=nil {
```
Chyba w "users"

shhhQuiettt replied:
Chyba nie pusty context.Background() tylko ten z bazy

akruszewski replied:
ctx powinien zostac przekazany do funkcji jako pierwszy parametr. Poczytaj: https://golang.org/pkg/context/

akruszewski commented on baza/baza.go:
```diff
@@ -0,0 +1,39 @@
+var Ctx, _ = context.WithTimeout(context.Background(), 10*time.Second)
+func connectToMongo () {//funkcja łączy się z bazą
+
+	client, err := mongo.NewClient(options.Client().ApplyURI("mongodb://127.0.0.1:27017"))
```
Ten url powinien isc z enva.

akruszewski commented on server.go:
```diff
@@ -11,5 +12,7 @@ func main() {
 		return c.SendString("Hello, World!")
 	})
 
+	app.Post("/newUser",user.SetUser)//endpoint dodawania uzytkownika do bazy
```
Najlepiej `createUser`, albo `registerUser`.

akruszewski commented on serviceDB/serviceDB.go:
```diff
@@ -0,0 +1,30 @@
+
+func InsertUser(user *person.Person ) error {//dodawanie użytkownika
+	if errv:=baza.DB.Collection("zadania").FindOne(context.Background(), bson.M{"email":user.Email}); errv !=nil {
+		return errors.New("konto istnieje")
```
cala komunikacja backendu najlepiej po angielsku, ewentualnie message bylyby tlumaczone na froncie.

akruszewski commented on user/user.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+func SetUser(c *fiber.Ctx) error {//dodawanie użytkownika do bazy
+	fmt.Println("jestem")
```
Do PR nie wypychamy komentow ktore nie powinny sie znalezc w produkcyjnym kodzie.

akruszewski commented on user/user.go:
```diff
@@ -0,0 +1,48 @@
+			"error": err.Error()})
+	}
+
+	fmt.Println(p.Name)
```
Do PR nie wypychamy komentow ktore nie powinny sie znalezc w produkcyjnym kodzie.

akruszewski commented:
Logowanie najlepiej ogarnij z logrus. wszystkie errory powinny byc logowane, dobrze tez logowac warningi i info.

wojbog committed:
Endpoint rejestracji - poprawa

naprawa context,pobieranie adresu, portu z  env. ,dodanie log
test validacji

akruszewski commented on baza/baza.go:
```diff
@@ -0,0 +1,36 @@
+
+func ConnectToMongo() { //funkcja łączy się z bazą
+
+	client, err := mongo.NewClient(options.Client().ApplyURI(os.Getenv("URL_BAZA")))
```
wyglada git, ale najlepiej nazwij to MONGO_URL

akruszewski commented on baza/baza.go:
```diff
@@ -0,0 +1,36 @@
+		log.Info("connect to DB")
+	}
+
+	DB = *client.Database("praktyki")
```
to tez dalbym do enva jako MONGO_DB, i jezeli nie ma takiej zmiennej, to ustawilbym default, np. "praktyki", ale moze lepiej nazwe aplikacji, czy cos w ten desen.

akruszewski commented on server.go:
```diff
@@ -1,15 +1,23 @@
 	app := fiber.New()
 
 	app.Get("/", func(c *fiber.Ctx) error {
 		return c.SendString("Hello, World!")
```
Tego bym sie pozbyl.

akruszewski replied:
Ponawiam request o zmiane.

akruszewski commented on user/user.go:
```diff
@@ -0,0 +1,46 @@
+	}
+
+	
+	log.Info("success")
```
Malo informatywny ten log. Jaka akcja skonczyla sie z sukcesem? Jaki np. user id zostal dodany, itp.

akruszewski reviewed (changes requested):
Juz jest prawie ok, ale te rzeczy z enva powinny byc sprawdzone czy istnieja, i jezeli nie, to powinna zostac podjeta jakas akcja: ustawienie wartosci defaultowej, albo np. zwrocenie bledu, jezeli zmienna musi byc ustawiona.

wojbog committed:
endpoint rejestracja - poprawa II

dodanie sprawdzania env

wojbog committed:
dodanie przypadkow

akruszewski commented on baza/baza.go:
```diff
@@ -0,0 +1,44 @@
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var DB mongo.Database
```
Globalne zmienne to na prawde zly pomysl, wszystko powinno byc przekazywane przez funkcje, i, jezeli istnieje taka potrzeba, trzymane w struktorach. Problem z globalsami jest taki, ze bardzo latwo mozesz miec errory z dostepami, jezeli dwie metody beda probowaly sie dostac do tego samego miejsca w pamieci.

akruszewski commented on baza/baza.go:
```diff
@@ -0,0 +1,44 @@
+
+}
+func init () {
+	connectToMongo()
```
Jak w komentarzu wyzej, init polaczenia do mongo to nienajlepsza idea.

akruszewski commented on person/person_test.go:
```diff
@@ -0,0 +1,94 @@
+	
+)
+
+func TestVlidatePasswordFalse(t *testing.T) {
```
Literowka, powinno byc Validate

akruszewski replied:
I w zasadzie to cos w desen TestValidatePasswordReturnFalseIfPasswordIsInvalid

akruszewski replied:
Tyczy sie to wszystkich testow.

akruszewski commented on serviceDB/serviceDB.go:
```diff
@@ -0,0 +1,35 @@
+package serviceDB
```
proponuje repository/user/user.go, komunikacja z baza najlepiej zeby lezala w repositories, a logika biznesowa w services, handlery http w handlers, a routing w server.

wojbog committed:
endpoint rejestracja - restrukturyzacja

podzielenie na handlers(handlery), server(routing), services(logika biznesowa),repository(komunikacja z bazą)

akruszewski commented on handlers/createUser.go:
```diff
@@ -0,0 +1,46 @@
+	"github.com/wojbog/praktyki_backend/repository/person"
+)
+
+func CreateUser(c *fiber.Ctx) error {//dodawanie użytkownika do bazy
```
Komentarze po angielsku poprosze, i nie tak dokumentuje sie funkcje: https://blog.golang.org/godoc

akruszewski commented on repository/user/user.go:
```diff
@@ -0,0 +1,72 @@
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type Baza struct {
```
Repository

wojbog committed:
Endpoint registration

bug fix

akruszewski commented on service/person.go:
```diff
@@ -0,0 +1,74 @@
+
+}
+//Validation validate Person struct
+func (p *Person) Validation() ([]string, error) {//walidacja struktury użytkownika
```
Nazwalbym to Validate, bo to akcja, Validation bardziej by sie nadawalo do np. tworzenia obiektu Validation.

akruszewski commented on handlers/createUser.go:
```diff
@@ -0,0 +1,46 @@
+			"list":    tab})
+	}
+	//add to datebase
+	if str,err:=col.InsertUser(c.Context(),p);err!=nil {//dodawanie do bazy
```
to i walidacja powinna byc w serwisie, nie w handlerze, i serwis uzyty tutaj.

wojbog committed:
add service

add service to communicate between handlers and repository

akruszewski commented on handlers/createUser.go:
```diff
@@ -0,0 +1,37 @@
+		log.Info(err.Error())
+		return c.Status(400).JSON(&fiber.Map{
+			"success": false,
+			"error":   err.Error()})
```
Do uzytkownika nie chcesz wysylac errora ktory zucila aplikacja w tym przypadku, to nie bedzie mowilo zwyklemu uzytkownikowi + to moze bys security issue. Tutaj lepiej wyswietlic customowy error: "Cannot read request", czy cos w ten desen.

akruszewski commented on handlers/createUser.go:
```diff
@@ -0,0 +1,37 @@
+	}
+	//add to datebase
+	if tab, err:=s.AddNewUser(c.Context(),p); err!=nil {
+		return c.Status(400).JSON(&fiber.Map{
```
Tutaj robisz zalozenie, ze dostaniesz tylko validation error. Trzeba by rozroznic validation errory od nieoczekiwanych errorow, tj. 500.

wojbog committed:
add models, 500 error

new instance UserResponse to inform client about user

wojbog committed:
correction
